src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java

```diff
@@ -485,12 +524,228 @@ public void createSnapshotRowIdsParquetFile(
   }
 
+  /**
+   * Create Snapshot by Asset First, create parquet file for root table, including only rows that
+   * match the root values Then, walk relationships defined by asset, creating a parquet file per
+   * relationship
+   *
+   * @param assetSpec Asset Specification used to define the root table and which tables and columns
+   *     to include in snapshot
+   * @param snapshotId UUID associated with destination snapshot
+   * @param datasetDataSourceName Data source associated with source data from dataset
+   * @param snapshotDataSourceName Data source associated with destination snapshot
+   * @param requestModel Specifies the root values which determines which rows to include in root
+   *     table snapshot
+   * @return tableRowCounts - hash map of snapshot table names and number of rows included in
+   *     snapshot
+   * @throws SQLException
+   */
+  public Map<String, Long> createSnapshotParquetFilesByAsset(
+      AssetSpecification assetSpec,
+      UUID snapshotId,
+      String datasetDataSourceName,
+      String snapshotDataSourceName,
+      SnapshotRequestAssetModel requestModel)
+      throws SQLException {
+    Map<String, Long> tableRowCounts = new HashMap<>();
+
+    // First handle root table
+    AssetTable rootTable = assetSpec.getRootTable();
+    String rootTableName = rootTable.getTable().getName();
+
+    // Get columns to include for root table
+    List<SynapseColumn> columns =
+        rootTable.getColumns().stream()
+            .map(c -> c.getDatasetColumn())
+            .collect(Collectors.toList())
+            .stream()
+            .map(Column::toSynapseColumn)
+            .collect(Collectors.toList());
+    ST sqlCreateSnapshotTableTemplate = new ST(createSnapshotTableByAssetTemplate);
+    ST queryTemplate =
+        generateSnapshotParquetCreateQuery(
+            sqlCreateSnapshotTableTemplate,
+            IngestUtils.getSourceDatasetParquetFilePath(rootTable.getTable().getName()),
+            rootTable.getTable().getName(),
+            snapshotId,
+            IngestUtils.getSnapshotSliceParquetFilePath(snapshotId, rootTableName, "root"),
+            datasetDataSourceName,
+            snapshotDataSourceName,
+            columns);
+
+    AssetColumn rootColumn = assetSpec.getRootColumn();
+    queryTemplate.add("rootColumn", rootColumn.getDatasetColumn().getName());
+    String query = queryTemplate.render();
+    MapSqlParameterSource params =
+        new MapSqlParameterSource().addValue("rootValues", requestModel.getRootValues());
+    int rows;
+    try {
+      rows = synapseJdbcTemplate.update(query, params);
+    } catch (DataAccessException ex) {
+      logger.error("Error while creating root table parquet file on snapshot by asset", ex);
+      throw new PdaoException(
+          "Unable to create parquet file for the root table. This is most likely because the source dataset table is empty.  See exception details if this does not appear to be the case. Exception cause: "
```

Comment:
I think you can just have the exception be the second parameter to `PdaoException` so that it keeps the exception object and we get the full stack trace

Reply:
Good point! Fixed!